src/main/java/org/jenkinsci/plugins/workflow/support/steps/ExecutorStepDynamicContext.java

```diff
@@ -90,12 +91,33 @@ void resume(StepContext context) throws Exception {
         if (executor != null) {
             throw new IllegalStateException("Already resumed");
         }
-        Queue.Item item = Queue.getInstance().schedule2(task, 0).getItem();
-        if (item == null) {
-            // TODO should also report when !ScheduleResult.created, since that is arguably an error
-            throw new IllegalStateException("queue refused " + task);
+        // If Jenkins restarts or crashes while we are waiting below, on the next startup the task may already be in
+        // the queue, in which case we should reuse it instead of scheduling a second task.
+        AtomicReference<Queue.Item> itemRef = new AtomicReference<>();
+        Queue.withLock(() -> {
+            for (Queue.Item item : Queue.getInstance().getItems()) {
+                if (item.task instanceof ExecutorStepExecution.PlaceholderTask) {
+                    ExecutorStepExecution.PlaceholderTask itemTask = (ExecutorStepExecution.PlaceholderTask) item.task;
+                    if (task.equals(itemTask)) {
+                        itemRef.set(item);
+                        break;
+                    }
+                }
+            }
+        });
+        // TODO: Do we need to check executor slots too in case there was a task in the queue but it has already started?
```

Comment:
:thinking: could be possible, in case some agents reattached before we got to this point.